session-helper/background-monitor.c

```diff
@@ -0,0 +1,426 @@
+      FlatpakInstance *inst = (FlatpakInstance *)g_object_get_data (G_OBJECT (n), "instance");
+      gboolean found = FALSE;
+      for (j = 0; j < apps->len && !found; j++)
+        {
+          FlatpakInstance *app = g_ptr_array_index (apps, j);
+          found = same_instance (app, inst);
+        }
+      if (!found)
+        {
+          notify_notification_close (n, NULL);
+          g_ptr_array_remove (notifications, n);
+        }
+    }
+  G_UNLOCK (notifications);
+}
+
+static void
+allow_app (NotifyNotification *notification,
+           char               *action,
+           gpointer            user_data)
+{
+  FlatpakInstance *instance = user_data;
+
+  g_debug ("Allowing app %s to run in background", flatpak_instance_get_app (instance));
+
+  set_permission (flatpak_instance_get_app (instance), YES);
+
+  remove_notification (notification);
+}
+
+static void
+forbid_app (NotifyNotification *notification,
+            char               *action,
+            gpointer            user_data)
+{
+  FlatpakInstance *instance = user_data;
+
+  g_debug ("Forbid app %s to run in background", flatpak_instance_get_app (instance));
+
+  set_permission (flatpak_instance_get_app (instance), NO);
+  kill_instance (instance);
+
+  remove_notification (notification);
+}
+
+static void
+ignore_app (NotifyNotification *notification,
+            char               *action,
+            gpointer            user_data)
+{
+  FlatpakInstance *instance = user_data;
+
+  g_debug ("Let app %s run in background", flatpak_instance_get_app (instance));
+
+  remove_notification (notification);
+}
+
+static void
+stop_app (NotifyNotification *notification,
+          char               *action,
+          gpointer            user_data)
+{
+  FlatpakInstance *instance = user_data;
+
+  g_debug ("Stop app %s", flatpak_instance_get_app (instance));
+
+  kill_instance (instance);
+
+  remove_notification (notification);
+}
+
+static void
+send_notification (FlatpakInstance *instance,
+                   Permission       permission)
+{
+  g_autofree char *name = flatpak_instance_get_display_name (instance);
+  const char *summary;
+  g_autofree char *body = NULL;
+  NotifyNotification *notification;
+
+  summary = _("Background activity");
+  body = g_strdup_printf (_("%s is running in the background."), name);
+  notification = notify_notification_new (summary, body, NULL);
+
+  if (permission == UNSET)
+    {
+      notify_notification_add_action (notification, "allow", _("Allow"), allow_app, g_object_ref (instance), g_object_unref);
+      notify_notification_add_action (notification, "forbid", _("Forbid"), forbid_app, g_object_ref (instance), g_object_unref);
+    }
+  else if (permission == ASK)
+    {
+      notify_notification_add_action (notification, "ignore", _("Ignore"), ignore_app, g_object_ref (instance), g_object_unref);
+      notify_notification_add_action (notification, "stop", _("Stop"), stop_app, g_object_ref (instance), g_object_unref);
```

Comment:
If permission is ask, should we not have a way to change it to yes permanently?

Reply:
The 'ask' think is a bit up-in-the-air. Designers don't like it. I would say if we keep the 'ask' permission, this level of fine-grained control should go in the applications panel in the control-center.